# Dimensionality reduction and PCA

## 1. Discuss curse of dimensionality

- The Curse of Dimensionality is too many features
- As the number of features increases the efficiency of the algorithm is generally decreases as the number of dimentions increases which ultimetly increases the possibility of possible answers
- simple examples is if we draw a circle of unit radious the points are in 2D space but if we draw a spear of same radious the space is 3D and there are more space covered
- It is also possible that all features are are not equally important to predicy classes

## 2. Discuss any 3 dimensionality reduction techniques

- The main dimensional reduction techniques are

- Principal component analysis (PCA)
- Linear Discriminatory Analysis (LDA)
- Singular Value Composition
- Isomap embedding
- Locally linear embedding (LLE)


## 3. Explain PCA in your own words

- Principal Component Analysis (PCA) is a statistical technique used for data reduction without losing its properties
- It describes the composition of variances and covariances through several linear combinations of the primary variables
- It is about obtaining a unique set of orthogonal axes where the data has the largest variance
- Its main aim is to overcome the dimensionality of the problem
- The reduction of dimensionality should be such that when dropping higher dimensions, the loss of data is minimum

## 4. Perform classification on wine dataset using any model of your choice using all the features

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets
wine = datasets.load_wine()

In [ ]:
wine

{'DESCR': '.. _wine_dataset:\n\nWine recognition dataset\n------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 178 (50 in each of three classes)\n    :Number of Attributes: 13 numeric, predictive attributes and the class\n    :Attribute Information:\n \t\t- Alcohol\n \t\t- Malic acid\n \t\t- Ash\n\t\t- Alcalinity of ash  \n \t\t- Magnesium\n\t\t- Total phenols\n \t\t- Flavanoids\n \t\t- Nonflavanoid phenols\n \t\t- Proanthocyanins\n\t\t- Color intensity\n \t\t- Hue\n \t\t- OD280/OD315 of diluted wines\n \t\t- Proline\n\n    - class:\n            - class_0\n            - class_1\n            - class_2\n\t\t\n    :Summary Statistics:\n    \n    ============================= ==== ===== ======= =====\n                                   Min   Max   Mean     SD\n    ============================= ==== ===== ======= =====\n    Alcohol:                      11.0  14.8    13.0   0.8\n    Malic Acid:                   0.74  5.80    2.34  1.12\n    Ash:           

In [ ]:
X = wine.data
y = wine.target
df_X = pd.DataFrame(X, columns= wine.feature_names)
df_X.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [ ]:
from sklearn.preprocessing import StandardScaler

df_X= StandardScaler().fit_transform(df_X)

In [ ]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(df_X,y,test_size=0.3,random_state=42)


In [ ]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error

xgb_model = xgb.XGBClassifier(random_state=42)

xgb_model.fit(x_train, y_train)

XGBClassifier(objective='multi:softprob', random_state=42)

In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

y_test_pred = xgb_model.predict(x_test)

print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[19  0  0]
 [ 0 21  0]
 [ 0  1 13]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.95      1.00      0.98        21
           2       1.00      0.93      0.96        14

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



## 5. Perform classification on wine dataset using PCA and taking all top components such that 90% variance is explained by them

In [ ]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 7)
principalComponents = pca.fit_transform(df_X)
Weightages=pca.explained_variance_ratio_
print(Weightages)
Weightages.sum()

[0.36198848 0.1920749  0.11123631 0.0706903  0.06563294 0.04935823
 0.04238679]


0.8933679539739376

In [ ]:
principalDf = pd.DataFrame(data = principalComponents, columns = ['PC1','PC2','PC3','PC4','PC5','PC6','PC7'])
principalDf.head()

PC1       PC2       PC3       PC4       PC5       PC6       PC7
0  3.316751 -1.443463 -0.165739 -0.215631  0.693043 -0.223880  0.596427
1  2.209465  0.333393 -2.026457 -0.291358 -0.257655 -0.927120  0.053776
2  2.516740 -1.031151  0.982819  0.724902 -0.251033  0.549276  0.424205
3  3.757066 -2.756372 -0.176192  0.567983 -0.311842  0.114431 -0.383337
4  1.008908 -0.869831  2.026688 -0.409766  0.298458 -0.406520  0.444074

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(principalDf,y,test_size=0.3,random_state=42)

xgb_model.fit(x_train, y_train)

y_test_pred = xgb_model.predict(x_test)

print(confusion_matrix(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

[[18  1  0]
 [ 0 21  0]
 [ 0  1 13]]
              precision    recall  f1-score   support

           0       1.00      0.95      0.97        19
           1       0.91      1.00      0.95        21
           2       1.00      0.93      0.96        14

    accuracy                           0.96        54
   macro avg       0.97      0.96      0.96        54
weighted avg       0.97      0.96      0.96        54



## 6. Discuss the difference in results vs difference in number of features used for classification.

Applied XGBoost Classifier in both case

Case 1: without PCA
- Features = 13
- Accuracy = 0.98

Case 2: with PCA
- Features = 7
- Accuracy = 0.96

- By reducing dimentions from 13 to 7 which is almost saving of 50% space for data and multiple reduction in computational reqirements
- Where as the Accuracy is reduced by only 0.02 from 0.98 to 0.96
- So the results are much similar even after the reduction in dimentions
